### PR TITLE
Non-blocking alter for vanilla merge tree

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implementation of "non-blocking" alter for StorageMergeTree

### CODE REVIEW

**Comment by alesapin** [comment]
Need to remove:
- [ ] AlterDataPartTransaction
- [ ] alterDataPart
- [ ] analyzeAlterConversions
- [ ] columns_lock from MergeTreeDataPart (finally)
- [ ] removeEmptyColumnsFromPart
Make code less clumsy.

**Comment by alesapin** [comment]
Do not merge.

**Comment by qoega** [comment]
Please check that changes in docs are not needed after this changes https://clickhouse.tech/docs/en/query_language/alter/#query_language_queries_alter

File: dbms/src/Storages/MergeTree/IMergedBlockOutputStream.cpp [hunk 1]
```diff
@@ -33,4 +38,78 @@ Block IMergedBlockOutputStream::getBlockAndPermute(const Block & block, const Na
...
+    }
+
+    /// Remove files on disk and checksums
+    for (const String & removed_file : remove_files)
+    {
+        if (checksums.files.count(removed_file))
+        {
+            String path_to_remove = data_part->getFullPath() + removed_file;
+
+            if (0 != unlink(path_to_remove.c_str()))
```
Comment (Jokser) [inline dbms/src/Storages/MergeTree/IMergedBlockOutputStream.cpp]: Should go through disk interface.
disk->remove(data_part->getFullRelativePath() + removed_file)

**Review by Jokser — CHANGES_REQUESTED** [review]

**Comment by alesapin** [comment]
test failure is ok, just need to fix the test.

File: dbms/src/Storages/AlterCommands.cpp [hunk 2]
```diff
@@ -189,7 +191,7 @@ std::optional<AlterCommand> AlterCommand::parse(const ASTAlterCommand * command_
     else if (command_ast->type == ASTAlterCommand::DROP_CONSTRAINT && !command_ast->partition)
     {
         if (command_ast->clear_column)
-            throw Exception(R"("ALTER TABLE table CLEAR COLUMN column" queries are not supported yet. Use "CLEAR COLUMN column IN PARTITION".)", ErrorCodes::NOT_IMPLEMENTED);
+            throw Exception(R"("ALTER TABLE table DROP CONSTRAINT constraint_name" queries are not supported yet. Use "DROP CONSTRAINT constraint_name IN PARTITION".)", ErrorCodes::NOT_IMPLEMENTED);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/AlterCommands.cpp]: Does not make sense.

File: dbms/src/Storages/StorageMergeTree.cpp [hunk 3]
```diff
@@ -564,6 +511,7 @@ void StorageMergeTree::loadMutations()
             {
                 MergeTreeMutationEntry entry(disk, path, it->name());
                 Int64 block_number = entry.block_number;
+                LOG_DEBUG(log, "Loading mutation:" << it->name() << " entry commands size " << entry.commands.size());
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMergeTree.cpp]: Whitespace.

**Review by alexey-milovidov — APPROVED** [review]
Awesome!

**Pull request merged**